### PR TITLE
Distorm3 Hook

### PR DESCRIPTION
**Pull request by Starwarsfan2099**
Hook to load the dll required by distorm3 (https://github.com/gdabah/distorm)

### CODE REVIEW

File: PyInstaller/hooks/hook-distorm3.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2005-2015, PyInstaller Development Team.
+#
+# Distributed under the terms of the GNU General Public License with exception
+# for distributing bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+#By Starwarsfan2099
```
Comment (matysek) [inline PyInstaller/hooks/hook-distorm3.py]: @Starwarsfan2099 please add your name to `./doc/CREDITS.rst`, to the `PyInstaller 3.1` section.

File: PyInstaller/hooks/hook-distorm3.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+from PyInstaller.compat import is_win, is_darwin
+from PyInstaller.utils.hooks import get_package_paths
+
+if is_win:
+    datas = [ ("C:\python27\Lib\site-packages\distorm3\distorm3.dll", ''), ]
+else:
+    raise ValueError('This distorm3 hook only works on Windows!')
```
Comment (matysek) [inline PyInstaller/hooks/hook-distorm3.py]: Please remove this `else` statement and add a comment that it was tested only on Windows.

File: PyInstaller/hooks/hook-distorm3.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# for distributing bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+#By Starwarsfan2099
+from PyInstaller.compat import is_win, is_darwin
+from PyInstaller.utils.hooks import get_package_paths
+
+if is_win:
+    datas = [ ("C:\python27\Lib\site-packages\distorm3\distorm3.dll", ''), ]
```
Comment (matysek) [inline PyInstaller/hooks/hook-distorm3.py]: Do not use this hardcoded path but rather function `collect_dynamic_libs`

**Comment by matysek** [comment]
This is a follow-up from issue #1730.

**Comment by Starwarsfan2099** [comment]
@matysek `binaries = collect_dynamic_libs('distorm3')` doesn't work, that is why its a hard coded path. I have tried `collect_dynamic_libs` and `collect_data_files`. Both compile but the program returns `ImportError: Error loading the diStorm dynamic library (or cannot load library into process) saample returned -1` The `datas =` does work though!? Any ideas on fixing this?

**Comment by matysek** [comment]
@Starwarsfan2099 probably using the `destdir` argument of `collect_dynamic_libs`

**Comment by Starwarsfan2099** [comment]
@matysek Using `collect_dynamic_libs` returns `datas = [ ("C:\python27\Lib\site-packages\distorm3\distorm3.dll", 'distorm3'), ]` and `datas = [ ("C:\python27\Lib\site-packages\distorm3\distorm3.dll", ''), ]` The second one works and the only difference is `'distorm3'` and `''`

**Comment by Starwarsfan2099** [comment]
@matysek, Ok, since `collect_dynamic_libs` wouldn't work, I tried something else:

`datas = [ ( os.path.join(site.getsitepackages()[1], 'distorm3\distorm3.dll'), ''), ]`

This isn't completely hard coded but finds the users site-packages directory, the only directory distorm3.dll will work in. Is this acceptable?

**Comment by matysek** [comment]
> @matysek, Ok, since collect_dynamic_libs wouldn't work, I tried something else:

@Starwarsfan2099 Have you tried the `destdir` argument of `collect_dynamic_libs`

**Comment by Starwarsfan2099** [comment]
@matysek I looked at the code for `collect_dynamic_libs` and I saw the `destdir` argument. Would example usage be like if I am building `sample.py` then I add `destdir=sample\dist`? Sorry for all the trouble.

**Comment by Starwarsfan2099** [comment]
Just tried with the `destdir` argument, still returns `Import Error: Error loading the distorm dynamic library`.

**Comment by matysek** [comment]
@Starwarsfan2099 Have you tried just `destdir=''`?

**Comment by Starwarsfan2099** [comment]
@matysek Yep, I tried it, same thing...

**Comment by matysek** [comment]
@Starwarsfan2099 Could you please fix the `collect_dynamic_libs` function like this?
```
--- a/PyInstaller/utils/hooks/__init__.py
+++ b/PyInstaller/utils/hooks/__init__.py
@@ -1048,7 +1048,7 @@ def collect_dynamic_libs(package, destdir=None):
                 # Produce the tuple
                 # (/abs/path/to/source/mod/submod/file.pyd,
                 #  mod/submod/file.pyd)
-                if destdir:
+                if destdir is None:
                     # Libraries will be put in the same directory.
                     dest = destdir
                 else:
```
The empty string should not be ignored but used as correct value. Sorry about this bug.

**Comment by Starwarsfan2099** [comment]
@matysek Fixed the bug, everything still compiles, but again the executable doesn't work. I'm not an expert in python, but I think I know the problem. After playing in the python terminal, I discovered a few things:

Here is the way that worked and didn't:

`datas = [ ("C:\python27\Lib\site-packages\distorm3\distorm3.dll", ''), ]`

`datas = collect_dynamic_libs`which returns `[ ("C:\python27\Lib\site-packages\distorm3\distorm3.dll",'distorm3'), ]`

The difference is `''` in the first one and `'distorm3'` in the second one. The first `datas = [...(etc)` works but the second `datas = collect_dynamic_libs("distorm3")` dosent. Why does `collect_dynamic_libraries` return the `distorm3` and why would it cause the problem?

**Comment by matysek** [comment]
@Starwarsfan2099 Thanks. I will look at it.

**Comment by matysek** [comment]
@Starwarsfan2099 Thanks for this hook and sorry about the confusion. The right argument was `destdir='.'`

**Comment by Starwarsfan2099** [comment]
@matysek Thanks for the help! Now I know what to do next time a create a hook. Is there a limit to how many hooks one person can create?(and submit)

**Comment by matysek** [comment]
@Starwarsfan2099 There is no limit, however those hooks should be useful for others users too.

**Pull request merged**